pkg/controller/workload/bpfcache/locality_cache.go

```diff
@@ -0,0 +1,154 @@
+package bpfcache
+
+import (
+	"kmesh.net/kmesh/api/v2/workloadapi"
+)
+
+const (
+	REGION    = 1 << iota // 000001
+	ZONE                  // 000010
+	SUBZONE               // 000100
+	NODENAME              // 001000
+	CLUSTERID             // 010000
+	NETWORK               // 100000
+)
+
+type localityInfo struct {
+	region    string // init from workload.GetLocality().GetRegion()
+	zone      string // init from workload.GetLocality().GetZone()
+	subZone   string // init from workload.GetLocality().GetSubZone()
+	nodeName  string // init from os.Getenv("NODE_NAME"), workload.GetNode()
+	clusterId string // init from workload.GetClusterId()
+	network   string // workload.GetNetwork()
+	mask      uint32 // mask
+}
+
+func Valid(s string) bool {
+	return s != ""
+}
+
+func (l *localityInfo) Set(s string, param uint32) {
+	if !Valid(s) {
+		return
+	}
+	switch param {
+	case REGION:
+		l.region = s
+	case ZONE:
+		l.zone = s
+	case SUBZONE:
+		l.subZone = s
+	case NODENAME:
+		l.nodeName = s
+	case CLUSTERID:
+		l.clusterId = s
+	case NETWORK:
+		l.network = s
+	}
+	l.mask |= param
+}
+
+func (l *localityInfo) Clear(param uint32) {
+	l.mask &= ^param
+}
+
+func (l *localityInfo) IsSet(param uint32) bool {
+	return l.mask&param != 0
+}
+
+type LocalityCache struct {
+	LbPolicy               uint32
+	localityInfo           localityInfo
+	LbStrictIndex          uint32 // for failover strict mode
+	isLocalityInfoSet      bool
+	RoutingPreference      []workloadapi.LoadBalancing_Scope
+	isRoutingPreferenceSet bool
+	workloadWaitQueue      map[*workloadapi.Workload]struct{}
+}
+
+func NewLocalityCache() *LocalityCache {
+	return &LocalityCache{
+		localityInfo:           localityInfo{},
+		isLocalityInfoSet:      false,
+		RoutingPreference:      make([]workloadapi.LoadBalancing_Scope, 0),
+		isRoutingPreferenceSet: false,
+		workloadWaitQueue:      make(map[*workloadapi.Workload]struct{}),
+	}
+}
+
+func (l *LocalityCache) SetLocality(nodeName, clusterId, network string, locality *workloadapi.Locality) {
+	// notice: nodeName should set by processor or os.Getenv("NODE_NAME"),
+	l.localityInfo.Set(nodeName, NODENAME)
+	l.localityInfo.Set(locality.GetRegion(), REGION)
+	l.localityInfo.Set(locality.GetSubzone(), SUBZONE)
+	l.localityInfo.Set(locality.GetZone(), ZONE)
+	l.localityInfo.Set(clusterId, CLUSTERID)
+	l.localityInfo.Set(network, NETWORK)
+
+	l.isLocalityInfoSet = true
+}
+
+func (l *LocalityCache) SetRoutingPreference(s []workloadapi.LoadBalancing_Scope) {
+	// notice: s should set by lb.GetRoutingPreference()
+	if len(s) > 0 {
+		l.RoutingPreference = s
+		l.LbStrictIndex = uint32(len(s))
+		l.isRoutingPreferenceSet = true
+	}
+}
+
+func (l *LocalityCache) CanLocalityLB() bool {
+	log.Debugf("isLocalityInfoSet: %#v, isRoutingPreferenceSet: %#v", l.isLocalityInfoSet, l.isRoutingPreferenceSet)
+	return l.isLocalityInfoSet && l.isRoutingPreferenceSet
+}
+
+func (l *LocalityCache) CalcuLocalityLBPrio(wl *workloadapi.Workload) uint32 {
```

Comment:
not sure i understand this: From my perspective, the workload priority has to do with the original node info.

If the node and the workload both reside in same RoutingPreference [], the workload has highest priority.

Reply:
Yes, in the earliest design version, it was designed this way. During a kmesh community meeting, you suggested considering the current design format. Given that the kmesh process exists on every node as a DaemonSet, it is possible to pre-calculate the priority of each workload within a service relative to that node on each node. This way, when subsequent traffic arrives and accesses the service, the backend to serve can be selected based on the pre-calculated priorities of the workloads for that service.